keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -244,7 +245,206 @@ There are likely others.
+are easy to implement and arguably more desirable --- provided there
+is some form of fairness within each priority level.  The current
+thinking is in that direction: use priorities, with simple equal
+fairness among some categories of requests in each priority level.
+There are published scheduling algorithms that provide fairness, and
+we hope to use/adapt one of them to apply independently within the
+confines of each priroity level.
+
+Another issue is whether to manage QPS or concurrency or what.
+Managing QPS leaps first to mind, perhaps because it is a simple
+concept and perhaps because it is familiar from the self-restraint
+that clients apply today.  But we want to also take service time into
+account; a request flow with longer service times should get less QPS
+because its requests are "heavier" --- they impose more load on the
+apiserver.  A natural way to do this is with an inverse linear
+relation.  For example, when two CPU-bound request flows are getting
+equal CPU from the apiserver, and the first flow's requests have a
+service time that is X times the service time of the second flow's
+requests, the first flow's QPS is 1/X of the second's.  This is
+exactly analogous to what happens in networking: if two flows are
+getting the same bandwidth, and one flow's packets are X times as big
+as the second's, then the first flow's packets per second rate is 1/X
+that of the second flow.  This inverse linear relation amounts to
+managing the product of QPS * service time.  That is equivalent to
+managing concurrency.  Managing concurrency is an obvious choice for
+memory, and we now see it is a good choice for CPU too.  This is also
+a convenient choice because it is what the max-in-flight handler is
+doing today, so we would be making a relatively modest extension to
+that hendler's conception.
+
+Compared to traditional scheduling problems, ours is harder because of
+the combination of these facts: (1) (unlike a router handling a
+packet) the apiserver does not know beforehand how long a request will
+take to serve nor how much memory it will consume, and (2) (unlike a
+CPU scheduler) the apiserver can not suspend and resume requests.
+Also, we are really loathe to abort a request once it has started
+being served.  We are leaning towards adapting well known and studied
+scheduling technique(s); but adaptation is a form of invention, and we
+have not converged yet on what to do here.
+
+Another issue is how to combine two goals: protection of CPU, and
+protection of memory.  A related issue is the fact that there are two
+stages of memory consumption: a request held in a queue holds some
+memory, and a request being served may use a lot more.  The current
+thinking seems to be focusing on using one QPS or concurrency limit on
+requests being served, on the expectation that this limit can be set
+to a value that provides reasonable protection for both CPU and memory
+without being too low for either.
```

Comment:
I think that "1 unit for every request" is not what we would like to have long-term.

But I do believe, that it's the right thing to do for the MVP - we will extend that concept in the next phase.

Reply:
I think we are in agreement here.